### PR TITLE
Adding unit test for mbedtls_x509write_csr_set_extension()

### LINKED ISSUE #4577: Test mbedtls_x509write_csr_set_extension
There are no unit tests for `mbedtls_x509write_csr_set_extension`.

### PR DESCRIPTION
**Pull request by valeriosetti**
This PR resolves #4577

## Status
Ready

## Requires Backporting
Yes, implemented by #6509

## Requires Changelog Entry
No, test only

## Migrations
No

## Additional comments
The newly generated CSR test file was obtained with this command
`openssl req -new -subj '/C=NL/O=PolarSSL/CN=PolarSSL Server 1' -key server1.key  -sha1 -addext "extendedKeyUsage=serverAuth" -out server1.req.sha1.ext`

### CODE REVIEW

**Comment by mpg** [comment]
Regarding the PR description:

- please use [github keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) to link the PR to the corresponding issue.
- requires backporting: I'd say yes, as a general rule we do backport test improvements, unless there's a specific reason - which I don't think is the case here.
- please also add "requires changelog entry: no - test only" or something to that effect. Currently we're doing a manual verification of missing ChangeLog entries before each release, which is very tedious. We have plans for a better system, but in the meantime, it really helps if each PR that doesn't have a ChangeLog entry explicitly states that this was intentional and not an omission. (I've created a PR to update the PR template: https://github.com/Mbed-TLS/mbedtls/pull/6465)

File: tests/data_files/server1.req.sha1.ext
Comment (mpg) [inline tests/data_files/server1.req.sha1.ext:1]: These days we use the `Makefile` in the same directory as a form of executable documentation for how these files were generated, please add an entry there. (We still commit the files, as not all commands are deterministics, and this particular `Makefile` is _not_ part of our build system or test system, it's really just executable documentation.)

(But thanks for documenting the command used in the PR description, in the old days we were not doing that and it only caused pain later. So, now we have this Makefile which is easier that finding the commit the file was introduced, and going to the PR from there.)

File: tests/suites/test_suite_x509write.data
Comment (mpg) [inline tests/suites/test_suite_x509write.data:54]: I'm slightly unhappy with the choice of SHA-1 here. Its use in certificate is deprecated for [strong reasons](https://shattered.io/) and is disabled in our default verification profile. Since we have so many existing test data using SHA-1, we're re-enabling it in tests, so what you did works. But still, in the future we might want to update our data to not rely on it any more, and that's going to be just a bit easier if in the meantime we avoid SHA-1 for new data.

So, while it's not a blocker, I'd be happier if the new file used SHA-256 instead.

**Review by mpg — CHANGES_REQUESTED** [review]
Looks pretty good to me. Just one point about documentation, and a suggestion for improvement.

File: tests/data_files/server1.req.sha1.ext
Comment (valeriosetti) [reply tests/data_files/server1.req.sha1.ext]: I did the change as requested but I have a question.
Basically I just followed the easiest path by adding a new line in the Makefile with the command line I used to generate the CSR. However I noticed that this is not very elegant since other CSRs are generated with an internal program (which is inside this repo).
Unfortunately this program does not support extensions flags so it must be modified/extended accordingly in order to do so. However, if this change is only required for a single test, I don't know if it's worth to add this change...

@mpg What do you suggest me to do?

Comment (mpg) [reply tests/data_files/server1.req.sha1.ext]: I think in this case we want the test file to be generated by another program that ours, as it gives stronger guarantees of correctness: if we generated the file using our own program then the test would only ensure that if our code is correct today, then it remains correct in the future, while if we generate it with openssl, the test says that if OpenSSL is correct, then so are we (today and in the future), which seems better to me. So I'd recommend you keep the line as it is but perhaps add a comment like "generate this with OpenSSL as a comparison point to test we're getting the same result".

**Review by mpg — APPROVED** [review]
Thanks for addressing my feedback, looks good to me now.

**Comment by mpg** [comment]
Looks like we got some CI failures. Quick note about the CI: we're using two systems:
- Travis, which only runs a small subset of our tests, so it's not enough, but it tends to finish quickly so it's good for quick feedback, and all results are public;
- Jenkins, which has the full set of tests. This is currently in the final phase of a migration between the old Internal instance (full results only accessible to Arm employees) and the new Open instance (all results publicly visible). The new instance is feature complete and runs the same set of tests as the old one, which we're only keeping around while we sort out some stability/performance issues with the new one. All this to say, clicking on "details" for Internal is expected to fail for you, but Open works and is enough.

Note about naming for Jenkins: `all_u16-xxx` means it's a component from `tests/scripts/all.sh` that was run in our Ubuntu 16.04 container ([here are](https://github.com/Mbed-TLS/mbedtls-test/tree/master/resources/docker_files) our Dockerfiles if you're curious), so you can reproduce locally by running `tests/scripts/all.sh xxx` (assuming you've got all the right tools installed, some components require specific cross toolchains, etc.)

Glancing at the results, you might have forgotten an `#if defined(RELEVANT_MACRO)` somewhere.

File: tests/suites/test_suite_x509write.function
Comment (mpg) [inline tests/suites/test_suite_x509write.function:97]: You probably meant `/* MBEDTLS_X509_CSR_WRITE_C */` here.

Comment (valeriosetti) [reply tests/suites/test_suite_x509write.function:97]: Ops, sorry! Yes, you're right

**Review by mpg — APPROVED** [review]
LGTM

**Comment by mpg** [comment]
Looks good to me. It's up to you when to start preparing the backport, but most people do it when the main PR has a green CI and an approval - or you can also wait for this PR to have its second approval. It's a compromise between latency and risk of having to rework.

File: tests/suites/test_suite_x509write.function
Comment (mprse) [inline tests/suites/test_suite_x509write.function:143]:
```suggestion
                            MBEDTLS_OID_SIZE( MBEDTLS_OID_SERVER_AUTH ) ) == 0 );
```

File: tests/suites/test_suite_x509write.function
Comment (mprse) [inline tests/suites/test_suite_x509write.function:82]: Can we use suitable macro or add comment explaining why 30 bytes is used?

File: tests/suites/test_suite_x509write.function
Comment (mprse) [inline tests/suites/test_suite_x509write.function:83]:
```suggestion
    unsigned char *p = buf + sizeof( buf );
```

**Review by mprse — COMMENTED** [review]
Looks good. Found only a few minor issues.

File: tests/suites/test_suite_x509write.function
Comment (mpg) [reply tests/suites/test_suite_x509write.function:143]: General note about coding style: we'll soon be switching to a more standard style that will be automatically enforced, so I think until then we can be a bit more relaxed about it. I mean, it's good to try and keep it consistent it we can, but I don't think it would be worth reworking the PR if that was the only issue. (But there's a more important one with the magic number for a buffer size, thanks for catching it btw, which requires rework anyway, so, we can indeed fix the style while at it.)

File: tests/suites/test_suite_x509write.function
Comment (valeriosetti) [reply tests/suites/test_suite_x509write.function:82]: You're right, sorry for this.
TBH I set it empirically during the development, but now I added:

- a comment explaining the size of this buffer
- set the size accordingly (I also verified manually that this is the correct value by trying different builds)

Comment (valeriosetti) [reply tests/suites/test_suite_x509write.function]: @mpg @mprse 
This change also impacts PR #6509. Once we're OK with this one, I'll apply the same changes to than one

**Review by mpg — DISMISSED** [review]
LGTM. Thanks for documenting the buffer's size.

File: tests/suites/test_suite_x509write.function
Comment (mprse) [inline tests/suites/test_suite_x509write.function:96]: typo: 
```suggestion
 * to set an overall upper boundary which is always guaranteed.
```

**Review by mprse — COMMENTED** [review]
Thanks for adding the comment. Found only one typo in the comment.
Otherwise looks good.

**Review by mprse — APPROVED** [review]
LGTM

**Review by mpg — APPROVED** [review]

**Pull request merged**
